### ISSUE TITLE
Export YOLOv8 by using TensorRT with INT8 calibration

### ISSUE THREAD

**Issue by pornpra** [ultralytics/ultralytics#11848]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi

I have already trained custom YOLOv8 (FP32) and I have some questions about export it by using TensorRT with INT8 calibration. 

1. What is the minimum number of calibration datasets I should use during the calibration process with YOLOv8 to prevent a significant drop in accuracy?
2. How can I adjust the code below to use my custom calibration dataset for exporting to INT8?

```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
model.export(
    format="engine",
    dynamic=True, 
    batch=8, 
    workspace=4, 
    int8=True,
    data="coco.yaml", 
)

model = YOLO("yolov8n.engine", task="detect") # load the model
```

**Comment by glenn-jocher** [comment]
Hi there! Here's how you can handle exporting your YOLOv8 model to TensorRT with INT8 calibration:

1. **Calibration Dataset Size:** Generally, for INT8 calibration, using at least 1000 images from your dataset is advised to minimize any significant drop in accuracy. More would be better to represent the dataset's variability effectively.

2. **Custom Calibration Dataset:** You can specify your custom dataset during the export process by modifying the `data` argument in your code. Make sure this points to your dataset configuration file, similar to how `coco.yaml` is pointed out here:

```python
from ultralytics import YOLO

model = YOLO("path/to/your/custom_yolov8.pt") # ensure this points to your trained model
model.export(
    format="engine",
    dynamic=True,
    batch=8,
    workspace=4,
    int8=True,
    data="path/to/your/custom_dataset.yaml", # specify your dataset configuration here
)

model = YOLO("yolov8n.engine", task="detect") # load the TensorRT model
```
Just replace `"path/to/your/custom_dataset.yaml"` with the path to your dataset configuration file. This customization will enable the use of your intended dataset during the calibration process.

Happy exporting! 😊

**Comment by pornpra** [comment]
@glenn-jocher Thanks for your response. I have two more questions about INT8 calibration:

1. Can you share the file coco.yaml, or suggest where I can download it? I need to see the details inside this file to build my custom_dataset.yaml. Currently, I have many images (all in .jpg format) for the calibration process in my directory, but I'm unsure how to use them to construct custom_dataset.yaml.

2. Can I use the latest Ultralytics version for exporting to INT8 and then use it (engine format) in an older version? Is it supported?

**Comment by glenn-jocher** [comment]
Hi! Great to see your interest in diving deeper into INT8 calibration with YOLOv8! 😊

1. **coco.yaml file:** You can find the `coco.yaml` file in the Ultralytics GitHub repository under the `data` directory. Here is how a typical `coco.yaml` looks like, customize it according to your dataset structure:

```yaml
train: ../coco/train2017  # 118k images
val: ../coco/val2017  # 5k images
nc: 80
names: [ 'person', 'bicycle', ...]  # Class names
```

For your `custom_dataset.yaml`, ensure you specify the paths to your training and validation data accordingly, set `nc` (number of classes), and define your class `names`.

2. **Version Compatibility:** Generally, it's advisable to use the same version of Ultralytics for both exporting and deploying your models. While backward compatibility might exist in some scenarios, using different versions can sometimes lead to unpredicted issues due to differences in implementations and dependencies.

I hope this helps! Let us know if there's anything else you need. 🌟

**Comment by pornpra** [comment]
@glenn-jocher 

Can I use only the validation set for calibration, or do I need to use both the training set and a calibration set?

**Comment by glenn-jocher** [comment]
Hi! For INT8 calibration, using just the validation set can be sufficient, especially if it's representative of the kind of data the model will encounter in deployment. However, including diverse images from the training set can further enhance the robustness of your calibrated model. Just ensure that the chosen images are varied enough to cover different scenarios. Happy calibrating! 😊

**Comment by pornpra** [comment]
@glenn-jocher 
I have successfully converted YOLOv8s from FP32 to INT8 without any errors, thanks for your support. Next, I use the following Python code to test the performance of INT8 in terms of inference time and RAM usage, comparing it with YOLOv8s in FP32 (.pt) and FP16 (.engine) formats, using a single image.

My environment

- NVIDIA Jetson Orin Nano (8GB)
- Python 3.8.10
- Ultralytics 8.2.13

```python
import torch
from memory_profiler import memory_usage
from ultralytics import YOLO

device = "cuda" if torch.cuda.is_available() else "cpu"
source = '/home/frame446.jpg'

# Function to measure memory and run inference
def measure_memory_and_time():
    # Add necessary imports, variables, or setups here if any
    start_time = time.time()
    model = YOLO("/home/int8/yolov8s_int8.engine", task="detect")
    results = model.predict(source, device=device)
    end_time = time.time()
    inference_time = end_time - start_time
    print(f"Inference Time: {inference_time} seconds")
    return results, inference_time

# Measure initial memory usage (assumes memory_usage function is imported)
initial_memory = memory_usage(-1, interval=0.05, timeout=1)

# Run the function and measure memory during execution
mem_usage = memory_usage(
    (measure_memory_and_time, (), {}), 
    interval=0.05, 
    include_children=True,
    max_usage=False,
    retval=False
)

# Measure final peak memory usage
final_peak_memory = max(mem_usage)

# Calculate average memory usage
average_memory = sum(mem_usage) / len(mem_usage)

print(f"Peak Memory Usage: {final_peak_memory} MiB")
print(f"Average Memory Usage: {average_memory} MiB")
```
These are my test results

**FP32**

image 1/1 /home/frame446.jpg: 384x640 4 cars, 290.5ms
Speed: 8.8ms preprocess, 290.5ms inference, 420.8ms postprocess per image at shape (1, 3, 384, 640)
Inference Time: 6.397905588150024 seconds
Peak Memory Usage: 3443.51953125 MiB
Average Memory Usage: 2141.6328519570707 MiB

**FP16**

image 1/1 /home/frame446.jpg: 640x640 4 cars, 23.5ms
Speed: 8.9ms preprocess, 23.5ms inference, 444.8ms postprocess per image at shape (1, 3, 640, 640)
Inference Time: 5.507491827011108 seconds
Peak Memory Usage: 3884.09375 MiB
Average Memory Usage: 2270.1096929505816 MiB

**INT8**

image 1/1 /home/frame446.jpg: 640x640 2 cars, 12.1ms
Speed: 17.8ms preprocess, 12.1ms inference, 578.6ms postprocess per image at shape (1, 3, 640, 640)
Inference Time: 7.327944040298462 seconds
Peak Memory Usage: 3619.57421875 MiB
Average Memory Usage: 2443.6972307477677 MiB

My new questions are

Why does INT8 consume the highest average RAM usage compared to FP32 and FP16? Doesn't it reduce RAM usage during inference?

**Comment by LopezJER** [comment]
When calibrating with a custom dataset, which subset does YOLOv8 use in the custom_dataset.yaml: train, val, or test? Or does it merge these three subsets as a whole during calibration?

**Comment by poseidon2060** [comment]
> @glenn-jocher I have successfully converted YOLOv8s from FP32 to INT8 without any errors, thanks for your support. Next, I use the following Python code to test the performance of INT8 in terms of inference time and RAM usage, comparing it with YOLOv8s in FP32 (.pt) and FP16 (.engine) formats, using a single image.
> 
> My environment
> 
> * NVIDIA Jetson Orin Nano (8GB)
> * Python 3.8.10
> * Ultralytics 8.2.13
> * TensorRT 8.5.2.2
> 
> ```python
> import torch
> from memory_profiler import memory_usage
> from ultralytics import YOLO
> 
> device = "cuda" if torch.cuda.is_available() else "cpu"
> source = '/home/frame446.jpg'
> 
> # Function to measure memory and run inference
> def measure_memory_and_time():
>     # Add necessary imports, variables, or setups here if any
>     start_time = time.time()
>     model = YOLO("/home/int8/yolov8s_int8.engine", task="detect")
>     results = model.predict(source, device=device)
>     end_time = time.time()
>     inference_time = end_time - start_time
>     print(f"Inference Time: {inference_time} seconds")
>     return results, inference_time
> 
> # Measure initial memory usage (assumes memory_usage function is imported)
> initial_memory = memory_usage(-1, interval=0.05, timeout=1)
> 
> # Run the function and measure memory during execution
> mem_usage = memory_usage(
>     (measure_memory_and_time, (), {}), 
>     interval=0.05, 
>     include_children=True,
>     max_usage=False,
>     retval=False
> )
> 
> # Measure final peak memory usage
> final_peak_memory = max(mem_usage)
> 
> # Calculate average memory usage
> average_memory = sum(mem_usage) / len(mem_usage)
> 
> print(f"Peak Memory Usage: {final_peak_memory} MiB")
> print(f"Average Memory Usage: {average_memory} MiB")
> ```
> 
> These are my test results
> 
> **FP32**
> 
> image 1/1 /home/frame446.jpg: 384x640 4 cars, 290.5ms Speed: 8.8ms preprocess, 290.5ms inference, 420.8ms postprocess per image at shape (1, 3, 384, 640) Inference Time: 6.397905588150024 seconds Peak Memory Usage: 3443.51953125 MiB Average Memory Usage: 2141.6328519570707 MiB
> 
> **FP16**
> 
> image 1/1 /home/frame446.jpg: 640x640 4 cars, 23.5ms Speed: 8.9ms preprocess, 23.5ms inference, 444.8ms postprocess per image at shape (1, 3, 640, 640) Inference Time: 5.507491827011108 seconds Peak Memory Usage: 3884.09375 MiB Average Memory Usage: 2270.1096929505816 MiB
> 
> **INT8**
> 
> image 1/1 /home/frame446.jpg: 640x640 2 cars, 12.1ms Speed: 17.8ms preprocess, 12.1ms inference, 578.6ms postprocess per image at shape (1, 3, 640, 640) Inference Time: 7.327944040298462 seconds Peak Memory Usage: 3619.57421875 MiB Average Memory Usage: 2443.6972307477677 MiB
> 
> My new questions are
> 
> Why does INT8 consume the highest average RAM usage compared to FP32 and FP16? Doesn't it reduce RAM usage during inference?

hey @pornpra 
How did you convert the model to int16?
I found the code to convert to int8 but cannot figure out how to convert it to int16
Can you please share the code snipped that you used

**Comment by glenn-jocher** [comment]
INT8 models can sometimes use more RAM due to additional overhead from quantization tables and TensorRT's memory management. For more details, please refer to the TensorRT documentation.

**Comment by Petros626** [comment]
@pornpra also interested in this here: https://github.com/ultralytics/ultralytics/issues/11848#issuecomment-2325675526.

Did you make progress @poseidon2060 ?

**Comment by Petros626** [comment]
@glenn-jocher how to use training data for calibration?

**Comment by glenn-jocher** [comment]
To use training data for INT8 calibration, modify your `data.yaml` to point to your training images path and use `data="data.yaml"` during export. See the [TensorRT INT8 calibration documentation](https://docs.ultralytics.com/integrations/tensorrt/#configuring-int8-export) for details on dataset configuration.

**Comment by Petros626** [comment]
I think directly specifying the folder with validation (recommended,bc inference data) or training data works too.

**Comment by glenn-jocher** [comment]
You're correct that using validation data (recommended) or training data via a properly configured `data.yaml` file is the correct approach for INT8 calibration. The key requirement is that the YAML file points to representative calibration images, which can indeed be your validation set. For implementation details, see the [TensorRT INT8 calibration documentation](https://docs.ultralytics.com/integrations/tensorrt/#configuring-int8-export) which shows the required YAML structure.

**Comment by karthikiitm87** [comment]
I have trained the ultralytics yolo model with swin backbone (https://github.com/Y-T-G/community/blob/main/cfg/detect/swin_t.yaml) and have been trying to do int8 quantization following the above comments but I have been getting the error and unable to resolve it. Can anyone please help me understand where I am goign wrong? 

The following is my code:
# Training:
model  = YOLO('swin_t.yaml')
train_results = model.train(
    data='/home/karthiks/novator/notebooks/QAT_demo/gwhd2021.yaml',
    epochs=25,
    batch=16,
    device='cuda',
    workers=8,
    project='runs/train',
    name='swin_t')


# Model export
from ultralytics import YOLO
model = YOLO('runs/train/swin_t/weights/best.pt')
model.export(format='engine', dynamic=True, batch=1, int8=True, data='/home/karthiks/novator/notebooks/QAT_demo/gwhd2021.yaml')

The following is the error I am getting and I am unsure where I am going wrong. Is it that the swin_t backbone that is used is not correctly implemented for quantization?

:ERROR ❌ TensorRT: export failure 83.0s: 'NoneType' object does not support the context manager protocol
19:39] [TRT] [E] IBuilder::buildSerializedNetwork: Error Code 10: Internal Error (Could not find any implementation for node {ForeignNode[Range_4620.../model.0/m.3/m.3.1/attn/Transpose_5 + /model.0/m.3/m.3.1/attn/Reshape_18]}.)

**Comment by Petros626** [comment]
Maybe specify the used workspace 'ram' for it, I exported always with `None`, so the `ram` size is selected automatically.

**Comment by glenn-jocher** [comment]
You're right about adjusting workspace parameters. The error with the Swin backbone is likely a compatibility issue with TensorRT - custom backbones like Swin transformers may contain operations not well supported by TensorRT's conversion process. 

In addition to trying different workspace values as suggested, I'd recommend:

1. First export to ONNX format and verify it loads correctly
2. Use a more standard backbone if TensorRT compatibility is critical for your deployment
3. Consider simplifying the model architecture if possible

TensorRT has better support for standard CNN architectures like CSP backbones that are used in default YOLO models, so custom architectures may require additional troubleshooting steps.
